cloudtrail: ability to run integration tests offline

goneri opened the pull request:
Depends-On: https://github.com/ansible/ansible-zuul-jobs/pull/1649

:warning: This PR will be rebased on the two PR above are merged.

cloudtrail test-suite requires too much extra privileges and we cannot run the Cloudtrail interation tests in the CI.
Instead we run the test-suite offline with a recording of the last API interactions. This is way happen when
we call `./run.sh`.

If you introduce a change that require an update of the recording, you should call the `./record.sh` script.
This will refresh the `recording.tar.gz` tarball with the last API interactions.

tremble commented on plugins/module_utils/botocore.py:
```diff
@@ -103,6 +103,8 @@ def _boto3_conn(conn_type=None, resource=None, region=None, endpoint=None, **par
         profile_name=profile,
     )
 
+    enable_placebo(session)
+
```
IMO this warrants a changelog entry (documentation would also be nice)

tremble commented on tests/integration/targets/cloudtrail/README.md:
```diff
@@ -0,0 +1,8 @@
+Instead we run the test-suite offline with a recording of the last API interactions. This is way happen when
+we call `./run.sh`.
+
+If you introduce a change that require an update of the recording, you should call the `./record.sh` script.
+This will refresh the `recording.tar.gz` tarball with the last API interactions.
```
Given these tests rely on recordings of 7 different modules I'm a little concerned how stable the tests will be.

goneri replied:
I understand your concerns. I would brand this a experimental at this stage. If we think this become a source of pain in the future I will be happy to remove this. But if it works well enough, this can safe us a lot of (CI) time.

tremble commented on tests/integration/targets/cloudtrail/main.yml:
```diff
@@ -0,0 +1,10 @@
+---
+# Beware: most of our tests here are run in parallel.
+# To add new tests you'll need to add a new host to the inventory and a matching
+# '{{ inventory_hostname }}'.yml file in roles/s3_bucket/tasks/
+
```
```suggestion
```

Looks like this was copy and paste from a set of "parallel" integration tests :)

tremble reviewed (commented):
In addition to the inline comments:

- meta/runtime.yml needs updating
- it would be good to add the matching community.aws PR at the same time.

alinabuzachis commented:
I also think we would like to keep the commits history of the module. This is what we did for the previous promotions if I am not mistaken. @tremble @jillr

tremble commented:
> I also think we would like to keep the commits history of the module. This is what we did for the previous promotions if I am not mistaken. @tremble @jillr

Good point - *yes please*

goneri commented:
> I also think we would like to keep the commits history of the module. This is what we did for the previous promotions if I am not mistaken. @tremble @jillr

@alinabuzachis I actually tried (with https://github.com/ansible-collections/el_grandiose_module_promoter) but I end up with a large series of commits and it was painful to go through. I will retry later.

alinabuzachis commented:
> > I also think we would like to keep the commits history of the module. This is what we did for the previous promotions if I am not mistaken. @tremble @jillr
> 
> @alinabuzachis I actually tried (with https://github.com/ansible-collections/el_grandiose_module_promoter) but I end up with a large series of commits and it was painful to go through. I will retry later.

@goneri Yes, this also has been done with the promoter tool https://github.com/ansible-collections/amazon.aws/pull/410

goneri commented:
recheck

goneri commented:
recheck

goneri commented:
recheck

goneri commented:
@tremble do you know why the FIPS tests are failing?

```

TASK [ansibleawsmodule.client : Check that we connected to the FIPS endpoint] ***
task path: /home/zuul/.ansible/collections/ansible_collections/amazon/aws/tests/integration/targets/module_utils_core/roles/ansibleawsmodule.client/tasks/endpoints.yml:32
fatal: [localhost]: FAILED! => {
    "assertion": "\"ec2-fips:DescribeImages\" in fips_endpoint_result.resource_actions",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

jillr commented:
Git history should be retained when the migration happens, also the module should not be migrated in a PR that is making changes.  Modules will be migrated in a separate PR once they are all in a supportable state, and we've communicated with the community that we're ready to start moving code between repos.

goneri commented:
Agreed, I will prepare a 2nd PR with the test-suite related changes.

tremble commented:
@goneri The FIPS tests seem to be working fine - see #996 (only failure was the changelog test)

The issue appears to be related to this change *not* the tests.

tremble commented on plugins/module_utils/botocore.py:
```diff
@@ -164,7 +166,7 @@ def get_aws_connection_info(module, boto3=None):
     # access_key
 
-    endpoint_url = module.params.get('endpoint_url')
+    ec2_url = module.params.get('ec2_url')
```
```suggestion
    endpoint_url = module.params.get('endpoint_url')
```

tremble commented on plugins/module_utils/botocore.py:
```diff
@@ -185,11 +187,11 @@ def get_aws_connection_info(module, boto3=None):
         module.fail("Passing both a profile and access tokens is not supported.")
 
-    if not endpoint_url:
+    if not ec2_url:
```
```suggestion
    if not endpoint_url:
```

tremble commented on plugins/module_utils/botocore.py:
```diff
@@ -185,11 +187,11 @@ def get_aws_connection_info(module, boto3=None):
+    if not ec2_url:
         if 'AWS_URL' in os.environ:
-            endpoint_url = os.environ['AWS_URL']
+            ec2_url = os.environ['AWS_URL']
```
```suggestion
            endpoint_url = os.environ['AWS_URL']
```

tremble commented on plugins/module_utils/botocore.py:
```diff
@@ -185,11 +187,11 @@ def get_aws_connection_info(module, boto3=None):
+            ec2_url = os.environ['AWS_URL']
         elif 'EC2_URL' in os.environ:
-            endpoint_url = os.environ['EC2_URL']
+            ec2_url = os.environ['EC2_URL']
```
```suggestion
            endpoint_url = os.environ['EC2_URL']
```

tremble commented on plugins/module_utils/botocore.py:
```diff
@@ -248,7 +250,7 @@ def get_aws_connection_info(module, boto3=None):
             boto_params[param] = text_type(value, 'utf-8', 'strict')
 
-    return region, endpoint_url, boto_params
+    return region, ec2_url, boto_params
```
```suggestion
    return region, endpoint_url, boto_params
```

tremble reviewed (changes requested):
Bad rebase.

goneri commented:
> Bad rebase.

oh thank you @tremble.

goneri commented:
recheck

tremble commented:
It looks like this PR's now being restricted to just the tests for cloudtrail.

Being picky I'm going to suggest that we split it even further.

1) split off the module_util changes and test them separately
this means we can merge and review that code separately from the cloudtrail migrations, and while the cloudtrail migrations should be pretty much a rubber-stamp, I think having an MVP example of a test for the module_utils part of this change has some value, especially as I can imagine them being used for other modules (IAM...).  This may also have made the rebase mistake easier to spot.

2) cloudtrail integration tests

goneri commented:
> It looks like this PR's now being restricted to just the tests for cloudtrail.

I had to also push the content of https://github.com/ansible-collections/amazon.aws/pull/997. The #990 PR itself is this commit https://github.com/ansible-collections/amazon.aws/pull/990/commits/4106284a66a6d255921ff8534ff823fa3b4c9b7a

> Being picky I'm going to suggest that we split it even further.
> 
>     1. split off the module_util changes and test them separately
>        this means we can merge and review that code separately from the cloudtrail migrations, and while the cloudtrail migrations should be pretty much a rubber-stamp, I think having an MVP example of a test for the module_utils part of this change has some value, especially as I can imagine them being used for other modules (IAM...).  This may also have made the rebase mistake easier to spot.  (These **definitely** need a changelog entry)

I think we're on the same page actually. The PR looks messy right now because it also brings the content of #997. I will add a changelog entry for #990 today.

>     2. cloudtrail integration tests (no changelog needed)

This part will come with #997.

tremble commented:
Thanks for splitting this all up, I know I've been a little picky.

goneri commented:
> Thanks for splitting this all up, I know I've been a little picky.

No worries, it was the right thing to do.

goneri commented:
recheck
